### PR TITLE
Homework shared_ptr, weak_ptr, make_shared

### PR DESCRIPTION
**Pull request by mateusz-talma**

### CODE REVIEW

File: homework/shared_ptr/WeakPtr.hpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    ptr_ = nullptr;
+    cb_ = nullptr;
+}
+
+template <typename T>
+long WeakPtr<T>::use_count() const noexcept {
+    if (cb_) {
+        return cb_->getSharedCounter();
+    }
+    return 0l;
```
Comment (MaciejSalamonski) [inline homework/shared_ptr/WeakPtr.hpp:131]: What is 0l? It is a long number?

**Review by MaciejSalamonski — APPROVED** [review]
Good job :) I have only one question

File: homework/shared_ptr/WeakPtr.hpp (hunk 1 above)
Comment (mateusz-talma) [reply homework/shared_ptr/WeakPtr.hpp:131]: 0L is a long integer number with all bits set to zero. In this case, 0L is the same as 0 (0 will be cast to long)

Comment (MaciejSalamonski) [reply homework/shared_ptr/WeakPtr.hpp:131]: Thanks ;~}

File: exercises/FileOperations.cpp [hunk 2]
```diff
@@ -1,23 +1,65 @@
...
+    ~FileHandler() {
+        std::fclose(file_);
+=======
 int main()
 {
     FILE* fp = std::fopen("thisFileDoesNotExist.cpp", "r");
     if(!fp) {
         std::perror("File opening failed");
         return EXIT_FAILURE;
+>>>>>>> 9addd5891723a1d66eecbe4dfdc328007e9895c7
```
Comment (WBigaj) [inline exercises/FileOperations.cpp:31]: What are those merging leftovers?

File: exercises/FileOperations.cpp [hunk 3]
```diff
@@ -1,23 +1,65 @@
...
-    else if (std::feof(fp))
-        std::puts("End of file reached successfully");
+private:
+    FILE* file_ = nullptr;
+};
 
+<<<<<<< HEAD
+int main() {
+    try {
+    FileHandler fh("../ResourceD.cpp");
```
Comment (WBigaj) [inline exercises/FileOperations.cpp:57]: I think filepath can be constexpr variable.

File: homework/shared_ptr/SharedPtr.hpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        cb_->setDeleter([&]() { delete ptr_; });
+        ptr.ptr_ = nullptr;
+        ptr.cb_ = nullptr;
+    }
+    return *this;
+}
+
+template <typename T>
+void SharedPtr<T>::reset(T* ptr /* = nullptr*/,
+                         std::function<void(T*)> deleter /* = [](T* ptr = nullptr) { delete ptr; }*/) noexcept {
```
Comment (WBigaj) [inline homework/shared_ptr/SharedPtr.hpp:105]: What are those comment in the middle of the code?

File: exercises/FileOperations.cpp [hunk 5]
```diff
@@ -1,23 +1,65 @@
...
+    std::cout << fh;
+    } catch (FileOpeningError& foe) {
+        std::cerr << foe.what();
+    }
+=======
     std::fclose(fp);
+>>>>>>> 9addd5891723a1d66eecbe4dfdc328007e9895c7
```
Comment (WBigaj) [inline exercises/FileOperations.cpp:64]: Another merging leftovers here.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: exercises/FileOperations.cpp (hunk 3 above)
Comment (mateusz-talma) [reply exercises/FileOperations.cpp:47]: This is a file from another task, but actually, it should be constexpr :)

File: homework/shared_ptr/SharedPtr.hpp (hunk 4 above)
Comment (mateusz-talma) [reply homework/shared_ptr/SharedPtr.hpp:105]: We use this notation in my work team. This shows in the cpp file that the function argument can take its default value.
However, this solution has a disadvantage - this means duplication and will add the possibility of having the comment out of sync with the code (e.g. when we will change default value of argument in hpp file and we will not synchronize this changes in comment.

File: homework/shared_ptr/WeakPtr-ut.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+TEST(WeakPtrTest, WeakPtrMoveAssignment) {
+    cs::SharedPtr<int> sharedPtr{new int{WeakPtrTestConst::intValue}};
+    cs::WeakPtr<int> weakPtr{sharedPtr};
+    cs::WeakPtr<int> weakPtr2{};
+    weakPtr2 = std::move(weakPtr);
+    ASSERT_TRUE(weakPtr.expired());
+    ASSERT_FALSE(weakPtr2.expired());
+}
+
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr-ut.cpp:70]: copy assignment is not tested

**Review by ziobron — CHANGES_REQUESTED** [review]
exercises should not be commited. unique_ptr should not be commited.
make_shared is not working.

File: homework/shared_ptr/ControlBlock.hpp [hunk 7]
```diff
@@ -8,13 +8,19 @@ class ControlBlock {
...
-    void decreaseWeakRefCounter() { weakRefCounter_--; }
+    void decreaseWeakRefCounter() {
+        if (weakRefCounter_ > 0)
+            weakRefCounter_--;
+    }
     size_t getSharedCounter() { return sharedRefCounter_.load(); }
     size_t getWeakCounter() { return weakRefCounter_.load(); }
-    //void callDeleter() { deleter_(); }
-    void setDeleter(std::function<void()> del) { deleter_ = del; }
+    void setDeleter(std::function<void(T*)> del) { deleter_ = del; }
```
Comment (ziobron) [inline homework/shared_ptr/ControlBlock.hpp:22]: please set it only in the constructor and remove setDeleter function. We don't want to replace a deleter later. It should be bind to object during construction.

File: homework/shared_ptr/ControlBlock.hpp [hunk 8]
```diff
@@ -8,13 +8,19 @@ class ControlBlock {
...
+    virtual void callDeleter() {}
```
Comment (ziobron) [inline homework/shared_ptr/ControlBlock.hpp:23]: isn't it better to have it pure virtual here?

File: homework/shared_ptr/WeakPtr-ut.cpp [hunk 9]
```diff
@@ -68,3 +69,28 @@ TEST(WeakPtrTest, WeakPtrMoveAssignment) {
     ASSERT_FALSE(weakPtr2.expired());
 }
 
+TEST(WeakPtrTest, WeakPtrCopyAssignment) {
+    cs::SharedPtr<int> sharedPtr{new int{WeakPtrTestConst::intValue}};
+    cs::WeakPtr<int> weakPtr{sharedPtr};
+    cs::WeakPtr<int> weakPtr2 = weakPtr;
```
Comment (ziobron) [inline homework/shared_ptr/WeakPtr-ut.cpp:75]: copy constructor is used here. Use this form
```suggestion
    cs::WeakPtr<int> weakPtr2{some_other_shared_ptr};
    weakPtr2 = weakPtr;
```

**Review by ziobron — CHANGES_REQUESTED** [review]
almost good :)

**Review by ziobron — APPROVED** [review]
Well done 👍

**Comment by ziobron** [comment]
50 XP za implementację shared_ptra (3794)
25 XP za implementację weak_ptra (3795)
40 XP za implementację make_shared (3796)
{ "requirement_ids": [3794, 3795, 3796] }

**Pull request closed without merging**